python/tskit/metadata.py

```diff
@@ -120,6 +123,191 @@ def decode(self, data: bytes) -> bytes:
+
+        schema = enforce_fixed_properties(schema)
+        try:
+            StructCodecSchemaValidator.check_schema(schema)
+        except jsonschema.exceptions.SchemaError as ve:
+            raise exceptions.MetadataSchemaValidationError(str(ve)) from ve
+        self.schema = self.order_by_index(schema)
+
+        def make_encode(sub_schema):
+            """
+            Create a function that can encode objects of this schema
+            """
+            if sub_schema["type"] == "array":
+                element_encoder = make_encode(sub_schema["items"])
+                return lambda array: struct.pack("<Q", len(array)) + b"".join(
+                    element_encoder(ele) for ele in array
+                )
+            elif sub_schema["type"] == "object":
+                sub_encoders = {
+                    key: make_encode(prop)
+                    for key, prop in sub_schema["properties"].items()
+                }
+                return lambda obj: b"".join(
+                    sub_encoder(obj[key]) for key, sub_encoder in sub_encoders.items()
+                )
+            elif sub_schema["type"] == "string":
+                return lambda string: struct.pack(
+                    "<" + sub_schema["binaryFormat"], string.encode()
+                )
+            elif sub_schema["type"] == "null":
+                return lambda _: b""
+            else:
+                return struct.Struct("<" + sub_schema["binaryFormat"]).pack
+
+        def make_decode(sub_schema):
+            """
+            Create a function that can decode objects of this schema
+            """
+            if sub_schema["type"] == "array":
+                element_decoder = make_decode(sub_schema["items"])
+                array_length_size = struct.calcsize("<Q")
+
+                def array_decode(bytes_, offset):
+                    ret = []
+                    array_length = struct.unpack_from("<Q", bytes_, offset)[0]
+                    offset += array_length_size
+                    for _ in range(array_length):
+                        obj, offset = element_decoder(bytes_, offset)
+                        ret.append(obj)
+                    return ret, offset
+
+                return array_decode
+
+            elif sub_schema["type"] == "object":
+                sub_decoders = {
+                    key: make_decode(prop)
+                    for key, prop in sub_schema["properties"].items()
+                }
+
+                def obj_decode(bytes_, offset):
+                    ret = {}
+                    for key, sub_decoder in sub_decoders.items():
+                        obj, offset = sub_decoder(bytes_, offset)
+                        ret[key] = obj
+                    return ret, offset
+
+                return obj_decode
+
+            elif sub_schema["type"] == "string":
+                size = struct.calcsize("<" + sub_schema["binaryFormat"])
+                return lambda bytes_, offset: (
```

Comment:
These lambdas and private functions are getting a bit cryptic. Maybe we could make these into a set of small module level functions instead, like ``_struct_codec_decode_string`` or something? This would also make it easier to do a few small unit tests on them.

Reply:
Broken up and individually tested.